src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheContainer.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols;
+
+/// <summary>
+/// This type is synthesized to hold the cached delegates that target static method groups.
+/// </summary>
+internal sealed class DelegateCacheContainer : SynthesizedContainer
+{
+    private readonly Symbol _containingSymbol;
+    private readonly NamedTypeSymbol? _constructedContainer;
+    private readonly Dictionary<(TypeSymbol, MethodSymbol), FieldSymbol> _delegateFields = new();
+
+    /// <summary>Creates a type scoped concrete delegate cache container.</summary>
+    internal DelegateCacheContainer(TypeSymbol containingType, int generationOrdinal)
+        : base(GeneratedNames.DelegateCacheContainerType(generationOrdinal), containingMethod: null)
+    {
+        Debug.Assert(containingType.IsDefinition);
+
+        _containingSymbol = containingType;
+    }
+
+    /// <summary>Creates a method scoped generic delegate cache container.</summary>
+    internal DelegateCacheContainer(MethodSymbol ownerMethod, int topLevelMethodOrdinal, int ownerUniqueId, int generationOrdinal)
+        : base(GeneratedNames.DelegateCacheContainerType(generationOrdinal, ownerMethod.Name, topLevelMethodOrdinal, ownerUniqueId), ownerMethod)
```

Comment:
>GeneratedNames.DelegateCacheContainerType(generationOrdinal, ownerMethod.Name, topLevelMethodOrdinal, ownerUniqueId), ownerMethod [](http://example.com/codeflow?start=15&length=129)

I think we want to have a set of test targeted specifically at name ambiguity among owner methods. At the same level and at different levels.

Reply:
Yes, we have one(?) already I guess. So that's a good idea. Added some more in 35a9f64, please check.